### ISSUE TITLE
Use uintStorage as a storage for bytes32

### ISSUE THREAD

**Issue by k1rill-fedoseev** [poanetwork/tokenbridge-contracts#373]
In several places, the `bytesStorage` is used for storing `bytes32` variables, since there is no separate storage for `bytes32` types in `EternalStorage` contract.

Consider to use `uintStorage` as a storage for `bytes32`, in order to save some gas.

In particular, there are two functions in `MessageProcessor`, which are relying on this, `setFailedMessageDataHash` and `setTransactionHash`.
There is a third usage of this in `setNonce` of AMB Mediator, it was addressed in 4a615de

**Comment by k1rill-fedoseev** [comment]
There is nothing complicated with `setTransactionHash`, since `transactionHash` is a temporary variable, which do not persist between calls to bridge contract. So, this code part can be replaced with 
```solidity
function transactionHash() external view returns (bytes32) {
    return bytes32(uintStorage[TRANSACTION_HASH]);
}

function setTransactionHash(bytes32 _txHash) internal {
    uintStorage[TRANSACTION_HASH] = uint256(_txHash);
}
```
For `setFailedMessageDataHash` the situation is a little bit different, saved data is then being used again. I see two ways of dealing with it:
* Leave it as it is
* Support both ways of storing, something like:
```solidity
function failedMessageDataHash(bytes32 _txHash) external view returns (bytes32) {
    bytes32 key = keccak256(abi.encodePacked("failedMessageDataHash", _txHash));
    uint256 dataHash =  uintStorage[key];
    return dataHash > 0 ? bytes32(dataHash) : Bytes.bytesToBytes32(bytesStorage[key]);
}

function setFailedMessageDataHash(bytes32 _txHash, bytes data) internal {
    uintStorage[keccak256(abi.encodePacked("failedMessageDataHash", _txHash))] = uint256(keccak256(data));
}
```
For this case, I got the following numbers for execution costs:
* Old version with `bytesStorage`:
  * `setFailedMessageDataHash` is `42049` gas
  * `failedMessageDataHash` is `2521` gas
* New version with `uintStorage`:
  * `setFailedMessageDataHash` is `20740 ` gas
  * `failedMessageDataHash` is `1358` gas for reading new dataHash from `uintStorage`, `3418` gas for reading.

**Comment by akolotov** [comment]
Addressed in [`5.0.0`](https://github.com/poanetwork/tokenbridge-contracts/releases/tag/5.0.0)
